### DISCUSSION TITLE
Longhorn does not run inside the rancher/rancher container

### DISCUSSION THREAD

**Discussion by koraykutanoglu** [longhorn/longhorn#9050] [General]
## Describe the bug

I install Ubuntu 22.04 virtual machine and uninstall Rancher v2.8.3 on it. Then, when I remove longhorn in the local cluster, the pods do not revolt.

## To Reproduce

- Install your ubuntu 22.04 server.
- and I run the following commands respectively.

```
sudo apt-get update && apt-get upgrade -y; curl -fsSL https://get.docker.com -o get-docker.sh; sudo sh get-docker.sh
```
```
sudo apt-get update;sudo apt-get install open-iscsi;sudo systemctl enable --now iscsid;modprobe iscsi_tcp;echo "iscsi_tcp" >> /etc/initramfs-tools/modules;echo iscsi_tcp >/etc/modules-load.d/iscsi-tcp.conf;update-initramfs -u;systemctl restart iscsid;lsmod |grep iscsi
```
```
docker run -d --restart=unless-stopped \
  --name=rancher \
  -p 80:80 -p 443:443 \
  --privileged \
  rancher/rancher:v2.8.3
```

- After opening the rancher web interface, I go to the charts section and install longhorn directly without making any settings.
- and when I examine the pods, I see the following error in the logs.
```
longhorn spec: failed to generate spec: path "/var/lib/longhorn/" is mounted on "/" but it is not a shared mount
```

## Expected behavior

Direct neat installation of Longhorn.

## Environment

The most current version of longhorn, installed directly from the charts section, using the Ubuntu server installed with the Ubuntu Server 22.04 LTS ISO downloaded from the original site of Ubuntu, and the Rancher v2.8.3 version installed on it.

**Comment by ejweber** [comment]
If I am understanding you correctly, you are:

- SSHing into some Ubuntu VM (in a cloud provider, etc.).
- Running Rancher server inside a standalone Docker container by following [these instructions](https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/other-installation-methods/rancher-on-a-single-node-with-docker).
- Accessing the Rancher server UI.
- Attempting to install Longhorn on the local cluster.

If the above is correct, this is not a supported environment in which to attempt a Longhorn install. The local cluster (kubelet and all) for this setup runs inside Docker containers. It is similar to https://kind.sigs.k8s.io/ or https://k3d.io/v5.7.2/. There are other issues and questions discussing potential issues that can arise in those setups.

The immediate issue is that the longhorn-manager pod spec calls for the following bidirectional mount:

```
        - name: longhorn
          mountPath: /var/lib/longhorn/
          mountPropagation: Bidirectional
```

But /var/lib/longhorn in your install is just a directory on a file system that only exists inside a Docker container. And that file system, which has `private` propagation, cannot be remounted `shared` to the longhorn-pod. It is similar to https://github.com/longhorn/longhorn/issues/2402 and others.

```
eweber@laptop:~/longhorn-manager> docker ps
CONTAINER ID   IMAGE                    COMMAND           CREATED          STATUS          PORTS                                      NAMES
790bb00a15d7   rancher/rancher:latest   "entrypoint.sh"   10 minutes ago   Up 10 minutes   0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp   amazing_swanson

eweber@laptop:~/longhorn-manager> docker exec -it 790bb00a15d7 bash

790bb00a15d7:/var/lib/rancher # findmnt -o TARGET,PROPAGATION /
TARGET PROPAGATION
/      private,slave
```

```
CreateContainerError (failed to generate container "d5a0abe75ba708f6de740a987da4fac922a4309779860200833f19914c2e0933" spec: failed to generate spec: path "/var/lib/longhorn/" is mounted on "/" but it is not a shared mount)
```

It may be possible to overcome this issue by creating a bidirectional bind mount for `/var/lib/longhorn` when starting the Rancher container, but there is a good chance you will run into iSCSI issues down the line. See [this discussion](https://github.com/longhorn/longhorn/discussions/2702) for some idea of what might be next.

The normal way to use Rancher and Longhorn together is like this:

- Install Rancher server somewhere (the single node Docker installation is fine for non-production use).
- Use Rancher server UI to provision a downstream cluster on VMs or bare metal.
- Use the Rancher server UI to install Longhorn on the downstream cluster. (The downstream cluster is running kubelets in VMs / on bare metal nodes and interfacing with a container runtime in the expected way.

What is the use case for trying to run Longhorn within the single Docker container?
